### ISSUE TITLE
FSI cylinder with flag: Newton iteration did not converge.

### ISSUE THREAD

**Issue by marcfehling** [exadg/exadg#676]
I was playing with your examples on fluid structure interactions, specifically the Turek-Hron benchmark. With the [input.json](https://github.com/exadg/exadg/blob/master/applications/fluid_structure_interaction/cylinder_with_flag/input.json) and [input_aitken.json](https://github.com/exadg/exadg/blob/master/applications/fluid_structure_interaction/cylinder_with_flag/input_aitken.json) files, the application crashes.

This **only** happens when using a large number of MPI processes. The application finishes fine when using 1, 2, 4, and 8 MPI processes. For 12 MPI processes, I get the following exception at some advanced stage `>5300 time steps, t = 9.2842e+00`:
```
--------------------------------------------------------
An error occurred in line <116> of file <.../exadg/include/exadg/solvers_and_preconditioners/newton/newton_solver.h> in function
    std::tuple<unsigned int, unsigned int> ExaDG::Newton::Solver<VectorType, NonlinearOperator, LinearOperator, LinearSolver>::solve(VectorType&, const ExaDG::Newton::UpdateData&) [with VectorType = dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>; NonlinearOperator = ExaDG::Structure::ResidualOperator<2, double>; LinearOperator = ExaDG::Structure::LinearizedOperator<2, double>; LinearSolver = ExaDG::Krylov::SolverBase<dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> >]
The violated condition was: 
    norm_r_damp < (1.0 - tau * omega) * norm_r
Additional information: 
    Damped Newton iteration did not converge. Maximum number of iterations
    exceeded!

Stacktrace:
-----------
#0  .../build-exadg/libexadg.so: ExaDG::Newton::Solver<dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>, ExaDG::Structure::ResidualOperator<2, double>, ExaDG::Structure::LinearizedOperator<2, double>, ExaDG::Krylov::SolverBase<dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> > >::solve(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, ExaDG::Newton::UpdateData const&)
#1  .../build-exadg/libexadg.so: ExaDG::Structure::Operator<2, double>::solve_nonlinear(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&, double, double, double, bool) const
#2  .../build-exadg/libexadg.so: ExaDG::Structure::TimeIntGenAlpha<2, double>::do_timestep_solve()
#3  .../build-exadg/libexadg.so: ExaDG::TimeIntBase::advance_one_timestep_solve()
#4  .../build-exadg/libexadg.so: ExaDG::FSI::Driver<2, double>::apply_dirichlet_neumann_scheme(dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&, unsigned int) const
#5  .../build-exadg/libexadg.so: ExaDG::FSI::PartitionedSolver<2, double>::solve(std::function<void (dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>&, dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host> const&, unsigned int)> const&)
#6  .../build-exadg/libexadg.so: ExaDG::FSI::Driver<2, double>::solve() const
#7  ./solver: void ExaDG::run<2, double>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, ompi_communicator_t* const&, bool)
#8  ./solver: main
--------------------------------------------------------
```

This is the last generated output of the pressure. I used the same scale as the last output step that still looked okay.
![image](https://github.com/exadg/exadg/assets/18285973/7439eb83-0284-47f8-9563-090e3e39c874)

Can you confirm this? I am using the current deal.II master branch compiled with `gcc 14.1.1` and `openmpi 5.0.3`. I can also reproduce it with 10 MPI processes, but the pressure doesn't look as dramatic in that case.

**Comment by kronbichler** [comment]
Yes, I see the same error on my machine as well (clang-18 compiler, openmpi 4.1.6).

**Comment by richardschu** [comment]
I came back to check this issue this week. Sorry @marcfehling; I have the feeling that you wanted us to know, but it did not bother you further. Thank you for reporting, it is very much appreciated. I gave this issue a shot several times during the past 2 years it has been open, but never have I ever found anything substantial.

This time, I tested:

- vector norm on all processors before updating the interface data, always giving something consistent like:

<img width="359" height="96" alt="Image" src="https://github.com/user-attachments/assets/34e3011e-8dc0-47ca-bb3f-28af3f043893" />

(I think this checks only the locally owned part `vector.l2_norm()` on a `dealii::LA::distributed::Vector`, but I made sure to call `vector.update_ghost_values()` before)

- I filled the data to be exchanged with dummy data in `local_interpolate_stress_bc_boundary_face`:

<img width="796" height="979" alt="Image" src="https://github.com/user-attachments/assets/bb9a9b02-581c-4d36-a1c1-417d1ccc338a" />

and checked if the interpolated value on the other side (`NonLinearOperator::do_boundary_integral_continuous()`) is correct:

<img width="774" height="739" alt="Image" src="https://github.com/user-attachments/assets/b24870ce-8fa1-4215-8885-b128d65dd53b" />

and got something like `1e-5` as the relative error on a relatively coarse triangulation. This is done only with `double` it seems. Since there is also some interpolation happening, I think this value is fine. This value was also on the same order of magnitude for 1, 4, 12 cores.

All the interface data exchange uses the same classes (`ContainerInterfaceData` and `InterfaceCoupling`), so I only looked at fluid force -> structure Neumann condition here.

------------------------------------------------

So we know that the vectors are consistent across all ranks and the exchange of interface data works in parallel.

The only aspect were I found differences in behavior was when changing the number of cores was when also changing the tolerances of the linear, nonlinear and FSI solvers. Sometimes one would observe convergence for 1 core, but not for more, sometimes for 12 and 4, but not for 1. 

I think the default settings are chosen such that one can run these examples on a personal computer, so they are right on the line of being stable. The preconditioners used by default differ using a different number of cores. Working on the pressure wave benchmark, I could always find a setting that would work for a specific number of MPI ranks, but might or might not work for *any* number of MPI ranks used.

Ideally, one would use tight tolerances, but also using tight tolerances might not always help: a certain level of added mass effect will always be present, and I have the feeling that the numerical dissipation related to looser tolerances sometimes helps mitigate the added mass effect. The picture is overall blurry, but I could always find a setting for a specific number of MPI ranks that worked, if *any* number of MPI ranks worked ... which is not a proof of anything, but supports the theory above. 

I continue to have no solution for this problem, and somehow I am not convinced that the faulty behavior we see is related to an actual bug. I think it is the mix of methods we use combined with limited computing resources that lead us to choose default settings that are not the most robust ones.

**Comment by richardschu** [comment]
FYI @iprusak @gabcas98
Let me know if you feel otherwise and we can have a look together!

**Comment by iprusak** [comment]
> FYI [@iprusak](https://github.com/iprusak) [@gabcas98](https://github.com/gabcas98) Let me know if you feel otherwise and we can have a look together!

Thank you @richardschu! @gabcas98 and I are planning to start investigating the Turek-Hron benchmarks with our optimisation-based (de-)coupling very soon. Maybe we get a better idea where exactly the issue is. We'll let you know if something comes up!
